### DISCUSSION TITLE
Dashboard with multiple tabs (shiny.router), all tabs load on start

### DISCUSSION THREAD

**Discussion by shahreyar-abeer** [Appsilon/shiny.fluent#116] [Q&A] [answered]
In a `shiny.fluent` dashboard with multiple tabs (navigation handled with `shiny.router`), all the tabs load on start.
Thus slowing down the app.
Is there any solution to this?

**Comment by jakubsob** [comment, marked as answer]
Hi @shahreyar-abeer,

It's hard to tell what exactly causes your app to slow down without an example, but I've created a simple app that I think illustrates the issue. Let me know if this is what you're struggling with.
```r
library(shiny)
library(shiny.fluent)
library(shiny.router)

router <- make_router(
  route(
    "/",
    tagList(
      Label("Home page")
    )
  ),
  route(
    "hidden",
    tagList(
      Label("Hidden page"),
      textOutput("text")
    ),
    server = function(input, output, session) {
      output$text <- renderText({
        Sys.sleep(5)
        "Delayed text"
      })
    }
  )
)

ui <- fluentPage(router$ui)
server <- function(input, output, session) {
  router$server(input, output, session)
}

shinyApp(ui, server)
```
this app takes 5 seconds to open due to a `Sys.sleep` call. In order not to run server code on hidden pages we can introduce a reactive which tracks if given page has been opened:
```r
library(shiny)
library(shiny.fluent)
library(shiny.router)

router <- make_router(
  route(
    "/",
    tagList(
      Label("Home page")
    )
  ),
  route(
    "hidden",
    tagList(
      Label("Hidden page"),
      textOutput("text")
    ),
    server = function(input, output, session) {
      opened <- reactiveVal(FALSE)
      observe({
        # Set `opened` reactive to indicate whether this page has been opened
        # It runs only once, after page has been opened for the first time
        if (!opened()) {
          opened(shiny.router::get_page() == "hidden")
        }
      })

      output$text <- renderText({
        shiny::req(opened())
        Sys.sleep(5)
        "Delayed text"
      })
    }
  )
)

ui <- fluentPage(router$ui)
server <- function(input, output, session) {
  router$server(input, output, session)
}

shinyApp(ui, server)
```
The app opens instantly. When we go to `/hidden` page we wait 5 seconds for the text to appear. Then we can go to `/` and back to `/hidden` and the text is still there (we don't wait 5 seconds) as it hasn't been invalidated by `opened` (which is invalidated once in the observer).

Does this help you speed up your app?

**Reply by shahreyar-abeer** [reply]
Thanks a lot!
This is exactly what I required.

**Reply by rodrigoestrellac** [reply]
In case you have some reactive output that should not be calculated if the page is hidden what would be the best approach ?? 
For example, I have a reactive expression that is used to render a bar plot on one page and a scatter plot on another. If the active page is the one with the bar plot, and the reactive expression changes, only the bar plot should be recalculated.

I guess the quicker way would be to create a reactive expression like this:
`activePage <- reactive({
  shiny.router::get_page()
})`

And then every output expression should start with `req(activePage() == "pageName")`, where pageName is the name of the page where the output is placed.

In the example I gave it should look like this 
`output$barplot <- renderPlot({
  req(activePage() == "barplotPage")
  
  plot
})

output$scatter_plot <- renderPlot({
  req(activePage() == "scatter_plotPage")
  
  plot
})
`

Am I right ? Is there a more efficient way to achieve this behaviour ??

Thanks !
